packages/git/src/node/git-repository-locator.ts

```diff
@@ -4,32 +4,75 @@
 import { FileUri } from '@theia/core/lib/node/file-uri';
 import { Repository } from '../common/model';
 
+const ignore: () => IgnoreFilter = require('ignore');
 const abs: (path: string) => string = require('abs');
-const finder: any = require('findit2');
-
-/**
- * Resolves to an array of repositories, recursively discovered from the given root `path`.
- *
- * @param path the FS path of the root to start the discovery.
- */
-export function locateRepositories(path: string): Promise<Repository[]> {
-    return new Promise<Repository[]>((resolve, reject) => {
-        const repositoryPaths = new Set();
-        const emitter = finder(abs(path));
-        emitter.on('directory', (dir: string, stat: fs.Stats, stop: () => void) => {
-            const base = Path.basename(dir);
-            if (base === '.git') {
-                const dirName = Path.dirname(dir);
-                repositoryPaths.add(dirName);
-                stop();
-            }
+const finder = require('findit2');
+
+export interface IgnoreFilter {
+    add(patterns: string): void
+    ignores(pathname: string): boolean;
+}
+export type IngoreFilterFn = (pathname: string) => boolean;
+
+@injectable()
+export class GitRepositoryLocator {
+
+    /**
+     * Resolves to an array of repositories, recursively discovered from the given root `path`.
+     *
+     * @param path the FS path of the root to start the discovery.
+     */
+    locate(path: string): Promise<Repository[]> {
+        return new Promise<Repository[]>((resolve, reject) => {
+            const filters = new Map<string, (pathname: string) => boolean>();
+            const repositoryPaths = new Set();
+            const emitter = finder(abs(path));
+            emitter.on('directory', (dir: string, stat: fs.Stats, stop: () => void) => {
+                const base = paths.basename(dir);
+                if (base === '.git') {
+                    const dirName = paths.dirname(dir);
+                    repositoryPaths.add(dirName);
+                    stop();
+                    return;
+                }
+                if (this.shouldStop(dir, filters)) {
+                    stop();
+                    return;
+                }
+                filters.set(dir, this.createFilter(dir));
+            });
+            emitter.on('end', () => resolve([...repositoryPaths].map(p => <Repository>{ localUri: FileUri.create(p).toString() })));
+            emitter.on('error', (error: Error) => reject(error));
         });
-        emitter.on('end', () => resolve([...repositoryPaths].map(p => <Repository>{ localUri: FileUri.create(p).toString() })));
-        emitter.on('error', (error: Error) => reject(error));
-    });
+    }
+
+    protected createFilter(path: string): IngoreFilterFn {
+        if (fs.existsSync(paths.join(path, '.gitignore'))) {
+            const ig = ignore();
+            ig.add(fs.readFileSync(paths.join(path, '.gitignore')).toString());
+            return pathname => ig.ignores(pathname);
```

Comment:
If you want to comply with how git handles nested `.gitignore` we would need to compose them unless it is a nested repo `.git` found. I have that logic in `file-search`. I think we should provide a generic igitIgnore file traverse function, that we can reuse in both. The `ignore` looks more popular then the one I used, so maybe we should go with that. But since the perf of GIT is really bad in current master I would like to tackle this in a subsequent PR.

Reply:
`shouldStop` will recursively apply filters from all parents

Reply:
I agree with extracting the common service for git-ignore, but don't want to refactor it in this PR.

Reply:
I see my is bogus, the child gitignore can negate patterns from the parent gitignore

Reply:
So we agree :)

Reply:
> `shouldStop` will recursively apply filters from all parents

It should stop on the next parent git repo, though. But let's tackle that separately.

Reply:
Please have a look again, I've fixed all comments.